### PR TITLE
Overlord

### PR DESCRIPTION
**Pull request by hebelken**
Going to read up on how to better structure the language of my tests. I think that is the biggest shortcoming of this commit. Let me know if you have any pointers.

### CODE REVIEW

File: overlord/Gemfile [hunk 1]
```diff
@@ -4,3 +4,7 @@ source "https://rubygems.org"
 ruby "2.0.0"
```
Comment (enovajuan) [inline overlord/Gemfile]: You need to add rspec to your gemfile as well.

Comment (enovajuan) [inline overlord/Gemfile]: although I can install in manually, out of the box I'm unable to run test... which is bad. Try creating a blank ruby sdk (or a gemset), run bundle install, then try running your tests, you may find what your gemfile is missing.

**Comment by enovajuan** [comment]
Tests Pass which is good. UI is somewhat confusing. Uncertain whom I need to log on with (dev/villain/other). Possibly add a read me with instructions or change the UI based on which user is logged in, maybe? Or maybe changing the message response color, once a code has been entered/submitted/etc, between green and red, to give some visual indication of bomb being active, or not.

Like any good super-villain, imagine said super-villain is not tech savy, or rather use iOS with the Grandparents mode turned on, and allow the front-end UI to `guide` me through the process. If you have any questions, feel free to ping me on HipChat/email/come by LPH, etc.

**Comment by enovajuan** [comment]
Overall nice work! Something as simple as color change when the backend responds or some other UI effect will push it past the finish line.

File: overlord/bomb.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    @defuse_attempts = nil
+    @status = "Inactive"
+  end
+
+  def explode_bomb
+    @status = "Exploded"
+  end
+
+  def default_status
+    "Offline"
```
Comment (djkotowski) [inline overlord/bomb.rb]: This should be a constant.

File: overlord/bomb.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  def default_activation_code(code)
```
Comment (djkotowski) [inline overlord/bomb.rb]: Confusing naming on this and `default_deactivation_code`.  I'd expect a constant with the default code, not a method that chooses which code (default or custom) to use.

File: overlord/features/activate.feature [hunk 4]
```diff
@@ -0,0 +1,43 @@
+Feature: Activate
+  As a villain
+  I want to be able to activate the bomb
+  In order to intensify extortion
```
Comment (djkotowski) [inline overlord/features/activate.feature]: Extortion intensifying.

File: overlord/features/boot.feature [hunk 5]
```diff
@@ -0,0 +1,36 @@
+Feature: Boot
+  As a villain
+  I want to be able to boot the bomb
+  In order to begin extorting society
+
+  Scenario: Bomb offline until booted
+    When I login as a villain
```
Comment (djkotowski) [inline overlord/features/boot.feature]: Nitpick: spelling

File: overlord/features/login.feature [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  I should only allow known users to enter the site
+
+  Scenario: Redirect to Login
+    When I visit the bomb page
+    Then I should be redirected to the login page
+
+  Scenario Outline: Login redirects to bomb
+    Given I am on the login page
+    When I login as <username>
+    Then I should be redirected to the <redirect_page> page
```
Comment (djkotowski) [inline overlord/features/login.feature]: Just specify "bomb" here and remove the column in the examples.  DRY it up.

File: overlord/features/step_definitions/bomb_steps.rb [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  current_path.should eq '/'
+  page.should_not have_content('villain')
+end
+
+When(/^I visit the reset page$/) do
+  visit '/reset'
+end
+
+Given(/^the bomb has exploded$/) do
+  step 'I am logged in as a villain'
```
Comment (djkotowski) [inline overlord/features/step_definitions/bomb_steps.rb]: It's generally bad practice to refer to other steps in a step definition.  Feel free to make methods to encapsulate repeated logic.

File: overlord/features/step_definitions/bomb_steps.rb [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+When(/^I use the wrong activation code$/) do
+  fill_in('code', with: "4321")
+  click_button('Apply Code')
+  page.find("#confirm_apply")
+  click_button('Confirm Code')
+end
+
+Then(/^I should see an inactive bomb$/) do
+  page.should have_content('Inactive')
```
Comment (djkotowski) [inline overlord/features/step_definitions/bomb_steps.rb]: Use the newer `expect` syntax instead of `should`.

File: overlord/overlord.rb [hunk 9]
```diff
@@ -1,14 +1,96 @@
...
+
+def set_session
+  session[:bomb] = session[:bomb]
+end
+
+def check_status
+  loop do
+    set_session
+    redirect '/explosion' if session[:bomb].status == "Exploded"
+    sleep 1
```
Comment (djkotowski) [inline overlord/overlord.rb]: Why do you have a loop with `sleep` here?

File: overlord/overlord.rb [hunk 10]
```diff
@@ -1,14 +1,96 @@
...
+BOMB_ALREADY_BOOTED = "The bomb has already been booted"
+INVALID_ACTIVATION_CODE = "Activation codes must be numeric"
+INCORRECT_ACTIVATION_CODE = "Incorrect activation code"
+
+def flash_message(message)
+  flash[:error] = message
+end
+
+def set_session
+  session[:bomb] = session[:bomb]
```
Comment (djkotowski) [inline overlord/overlord.rb]: This does nothing.

File: overlord/overlord.rb [hunk 11]
```diff
@@ -1,14 +1,96 @@
...
+  flash_message(BOMB_NOT_BOOTED) unless session[:bomb].booted?
+end
+
+def validate_activation_code(code)
+  if code =~ /\A\d+\z/ || code == ''
+    true
+  else
+    flash_message(INVALID_ACTIVATION_CODE)
+    false
+  end
```
Comment (djkotowski) [inline overlord/overlord.rb]: Use guard clauses to simplify.

File: overlord/overlord.rb [hunk 12]
```diff
@@ -1,14 +1,96 @@
...
+  redirect '/' unless session[:username]
+  erb :bomb
+end
+
+post '/boot' do
+  is_valid = validate_activation_code(params[:activation_code])
+  args = {
+    activation_code: params[:activation_code],
+    deactivation_code: params[:deactivation_code],
+  }
```
Comment (djkotowski) [inline overlord/overlord.rb]: L69-73 could be extracted to methods, reducing the size of this block.

File: overlord/spec/bomb_spec.rb [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  activate(bomb)
+  DEFAULT_DEFUSE_ATTEMPTS.times { bomb.apply_code("ZXCVMNDFBLJADFBNAAFBADFBD") }
+end
+
+describe Bomb do
+  DEFAULT_STATUS = "Offline"
+  DEFAULT_BOOT_STATUS = "Inactive"
+  DEFAULT_ACTIVATION_CODE = "1234"
+  DEFAULT_DEACTIVATION_CODE = "0000"
+  DEFAULT_DEFUSE_ATTEMPTS = 3
```
Comment (djkotowski) [inline overlord/spec/bomb_spec.rb]: If you adjust the default values in Bomb to be constants, you can access them from here.

File: overlord/views/bomb.erb [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    <input id="apply_code" type="button" value="Apply Code">
+    <input id="confirm_apply" type="submit" value="Confirm Code">
+    <input id="cancel_apply" type="button" value="Cancel">
+  </form>
+
+  <a class="explodable" href="/logout">logout</a>
+</div>
+
+<script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
+<script type="text/javascript">
```
Comment (djkotowski) [inline overlord/views/bomb.erb]: Extract JS out to a separate file.

**Comment by djkotowski** [comment]
Good iteration!  Take a look at the comments, refactor and resubmit.  Talk to @enovajuan or myself if you have any questions.

**Pull request closed without merging**